pkg/controller/nodeipam/node_ipam_controller.go

```diff
@@ -0,0 +1,201 @@
+		Key:    algorithm.TaintNodeNotReady,
+		Effect: v1.TaintEffectNoExecute,
+	}
+)
+
+const (
+	// ipamResyncInterval is the amount of time between when the cloud and node
+	// CIDR range assignments are synchronized.
+	ipamResyncInterval = 30 * time.Second
+	// ipamMaxBackoff is the maximum backoff for retrying synchronization of a
+	// given in the error state.
+	ipamMaxBackoff = 10 * time.Second
+	// ipamInitialRetry is the initial retry interval for retrying synchronization of a
+	// given in the error state.
+	ipamInitialBackoff = 250 * time.Millisecond
+)
+
+// Controller is the controller that manages node ipam state.
+type Controller struct {
+	allocateNodeCIDRs bool
+	allocatorType     ipam.CIDRAllocatorType
+
+	cloud       cloudprovider.Interface
+	clusterCIDR *net.IPNet
+	serviceCIDR *net.IPNet
+	kubeClient  clientset.Interface
+	// Method for easy mocking in unittest.
+	lookupIP func(host string) ([]net.IP, error)
+
+	nodeLister         corelisters.NodeLister
+	nodeInformerSynced cache.InformerSynced
+
+	cidrAllocator ipam.CIDRAllocator
+
+	forcefullyDeletePod func(*v1.Pod) error
+}
+
+// NewNodeIpamController returns a new node controller to sync instances from cloudprovider.
+// This method returns an error if it is unable to initialize the CIDR bitmap with
+// podCIDRs it has already allocated to nodes. Since we don't allow podCIDR changes
+// currently, this should be handled as a fatal error.
+func NewNodeIpamController(
+	nodeInformer coreinformers.NodeInformer,
+	cloud cloudprovider.Interface,
+	kubeClient clientset.Interface,
+	clusterCIDR *net.IPNet,
+	serviceCIDR *net.IPNet,
+	nodeCIDRMaskSize int,
+	allocateNodeCIDRs bool,
+	allocatorType ipam.CIDRAllocatorType) (*Controller, error) {
+
+	if kubeClient == nil {
+		glog.Fatalf("kubeClient is nil when starting Controller")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+
+	glog.V(0).Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+		})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("node_ipam_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	if allocateNodeCIDRs {
+		if clusterCIDR == nil {
+			glog.Fatal("Controller: Must specify clusterCIDR if allocateNodeCIDRs == true.")
+		}
+		mask := clusterCIDR.Mask
+		if maskSize, _ := mask.Size(); maskSize > nodeCIDRMaskSize {
+			glog.Fatal("Controller: Invalid clusterCIDR, mask size of clusterCIDR must be less than nodeCIDRMaskSize.")
+		}
+	}
+
+	nc := &Controller{
+		cloud:             cloud,
+		kubeClient:        kubeClient,
+		lookupIP:          net.LookupIP,
+		clusterCIDR:       clusterCIDR,
+		serviceCIDR:       serviceCIDR,
+		allocateNodeCIDRs: allocateNodeCIDRs,
+		allocatorType:     allocatorType,
+	}
+
+	if nc.allocateNodeCIDRs {
```

Comment:
I'd move this if to the /cmd/kube-controller-manger code, and initialize this controller only if `allocateNodeCIDRs` is set.

Reply:
Given that this logic could be in either KCM or CCM I would rather keep this logic out of kube-controller-manager. I would like to make this a bit more generic but I'd rather do that in a different PR. The split is already large and I would rather minimize this PR to things directly about the split and fix other issues in follow up PRs.

Reply:
SG, please add TODO.

To clarify - I just meant moving the external if outside of this function, as if allocateNodeCIDRs == false, this whole controller is no-op.